### PR TITLE
SILKIT-1477 SystemController Utility: Use SignalHandler, better report state

### PR DESCRIPTION
**Pull request by AndreasRentschler**
## Subject

Improvements of SystemController utility

## Description

[SILKIT-1477](https://jira.vg.vector.int/browse/SILKIT-1477)
System controller blocks new participants after required participants have terminated

- SystemController did not consistently report about the system's shutdown/error state.
- Reconnecting a participant after the system went into Shutdown state made the SystemController appear to hang.
- Changed to SignalHandler approach to allow for implicit exit after shutdown (`std::cin.ignore()` cannot be canceled).
  SIgnalHandler had to be turned into a reusable module, previously owned by Registry utility.
- Bugfix: Warning on valid transition from Aborting to Shutdown
- RequestReply was ignored when barrier has not yet been passed. Now, a warning is emitted and the new RequestReply is executed.

## Instructions for review / testing

- SignalHandler, refactored for public use
- Changes in SystemStateTracker.cpp and LifecycleStates.cpp

## Developer checklist (address before review)

- [ ] Changelog.md updated
- [ ] Prepared update for depending repositories
- [ ] Documentation updated (public API changes only)
- [ ] API docstrings updated (public API changes only)
- [ ] Rebase &rarr; commit history clean
- [ ] Squash and merge &rarr; proper PR title

### CODE REVIEW

File: SilKit/source/util/SignalHandler.cpp [hunk 1]
```diff
@@ -188,7 +205,27 @@ class SignalMonitor
     int _signalNumber{-1};
 };
 
-void systemHandler(int sigNum)
+static inline void setSignalAction(int sigNum, __sighandler_t action)
+{
+    // Check current signal handler action to see if it's set to SIGNAL IGNORE
+    struct sigaction oldAction;
```
Comment (MariusBgm) [inline SilKit/source/util/SignalHandler.cpp:211]:
```suggestion
    struct sigaction oldAction{};
```
better default initialize variables on stack

File: SilKit/source/util/SignalHandler.cpp [hunk 2]
```diff
@@ -188,7 +205,27 @@ class SignalMonitor
...
+    sigaction(sigNum, NULL, &oldAction);
+    if (oldAction.sa_handler == SIG_IGN)
+    {
+        // A non-job-control shell wants us to ignore this kind of signal
+
+    }
+    // Set new signal handler action to what we want
+    struct sigaction newAction;
```
Comment (MariusBgm) [inline SilKit/source/util/SignalHandler.cpp:219]:
```suggestion
    struct sigaction newAction{};
```
definitely initialize this here, as there might be garbage on the stack

File: SilKit/source/util/SignalHandler.cpp [hunk 3]
```diff
@@ -188,7 +205,27 @@ class SignalMonitor
...
 
-void systemHandler(int sigNum)
+static inline void setSignalAction(int sigNum, __sighandler_t action)
+{
+    // Check current signal handler action to see if it's set to SIGNAL IGNORE
+    struct sigaction oldAction;
+    sigaction(sigNum, NULL, &oldAction);
+    if (oldAction.sa_handler == SIG_IGN)
+    {
+        // A non-job-control shell wants us to ignore this kind of signal
```
Comment (MariusBgm) [inline SilKit/source/util/SignalHandler.cpp:215]: should we bail out here? if the sa_handler is set to SIG_IGN, our parent process already set it?

```suggestion
        // A non-job-control shell wants us to ignore this kind of signal
        return;
```

**Review by MariusBgm — CHANGES_REQUESTED** [review]
Only looked at the SignalHandler.cpp: we should initialize the stack variables

File: SilKit/source/core/requests/procs/ParticipantReplies.cpp [hunk 4]
```diff
@@ -47,8 +47,9 @@ void ParticipantReplies::CallAfterAllParticipantsReplied(std::function<void()> c
 
         if (_barrierActive)
         {
-            Services::Logging::Debug(_participant->GetLogger(), "Only one barrier can be active at a time, ignoring.");
-            return;
```
Comment (KonradBkd) [inline SilKit/source/core/requests/procs/ParticipantReplies.cpp:51]: Rückbau

File: SilKit/source/services/orchestration/LifecycleStates.cpp [hunk 5]
```diff
@@ -708,15 +708,15 @@ void ShutdownState::ShutdownParticipant(std::string reason)
             if (success)
             {
                 _lifecycleManager->NotifyShutdownInConnection();
-                _lifecycleManager->GetService()->SetFinalStatePromise();
```
Comment (KonradBkd) [inline SilKit/source/services/orchestration/LifecycleStates.cpp:711]: Rückbau

**Review by KonradBkd — CHANGES_REQUESTED** [review]
The two critical changes should be reverted, otherwise ok👍

**Comment by AndreasRentschler** [comment]
Thanks for your reviews, all suggestions were implemented.

**Review by KonradBkd — APPROVED** [review]
LGTM

**Review by MariusBgm — APPROVED** [review]

**Pull request merged**